### ISSUE TITLE
issue about training

### ISSUE THREAD

**Issue by puddinord** [mtangemann/deepgazemr#1]
Thanks for your work! I've been working on image saliency prediction recently. A few days ago, I reproduced deepgazeII using pytorch. But I found the loss is difficult to converge. The loss function I use is the maximum likelihood probability over the fixation points. The code is as follow，fixation is a matrix with only 0 and 1.

`log_probability = net(img)`
`loss = - torch.sum(fixation * log_probability.double()) / fixation.shape[0]`

During the training, the loss has been fluctuating around 5000. Surprisingly, after a few epochs, the model can make a basically correct prediction. Have you ever encountered this kind of problem?

**Comment by mtangemann** [comment]
I personally did not encounter this problem so far. The fact that the model yields good predictions nevertheless hints at the loss being off by a (near) constant factor or offset.

From looking at your code I assume that `fixation` has shape `(H,W)` and is non-zero where a subject looked. So I think to normalize the loss term over fixations you actually have to sum the entries from that matrix, i.e.

```
log_probability = net(img)
loss = - torch.sum(fixation * log_probability.double()) / torch.sum(fixation)
```

Or in the case of a batch size > 1, you would have to first normalize over fixations per sample and then over samples, as the number of fixations typically varies per sample:

```
log_probability = net(img)  # assuming shape (B,H,W)
loss = - torch.mean(torch.sum(fixation * log_probability.double(), dim=(1,2)) / torch.sum(fixation, dim=(1,2)))
```

Could that be the problem?

**Comment by puddinord** [comment]
thanks for your speedy reply. I tried as you suggested. The loss of the model after random initialization at epoch 1 is about 11.5. Then I loaded the weight that I had previously trained on salicon at 100 epochs , and the loss value was about 10. Does this mean that loss  is close to the optimal value in the beginning?

here is the prediction of model after 100 epoch training. The image from top to bottom is prediction map，gt map, fixations.
The model did learn the saliency of image.
![image](https://user-images.githubusercontent.com/20411487/98324436-34685c00-2027-11eb-948e-5f4e27242fdd.png)

**Comment by mtangemann** [comment]
Yes those numbers have the right scale now. When interpreting those results you have to take into account two aspects:

- The information gain reported in papers is the log likelihood relative to a baseline model (center bias, or uniform distribution). E.g. for the uniform distribution the information is calculated as follows:
```python
log_probability = net(img)  # assuming shape (B,H,W)
log_likelihood = torch.mean(torch.sum(fixation * log_probability.double(), dim=(1,2)) / torch.sum(fixation, dim=(1,2)))
log_likelihood_uniform = torch.log(1 / (H*W))  # constant -> no mean over fixated positions necessary
information_gain = log_likelihood - log_likelihood_uniform
```
- When using the natural logarithm (i.e. `torch.log`), the unit of the information gain is nats, not bits. To calculate the information gain in bits you have to multiply by `log2(e)`.

As both the baseline performance and the factor to calculate bits from nats are constant, they don't affect the optimization. But they're needed in order to interprete the numbers. In your example, assuming an image resolution of 640x480 pixel, the information gain has increased during training from ~1.3bit to ~3.8bit (relative to uniform). Having 1.3bit after random initialization looks a little high, so probably the resolution I assumed is wrong, but the numbers are in the right ballpark.

**Comment by puddinord** [comment]
Sorry, I didn't see your reply this weekend. 😟 

1.  I don't quite understand what 1.3-3.8bit means. Is it equivalent to `log_likelihood / log_likelihood_uniform ` ?  
2. What I mean by random initialization is only randomly initializing 1x1 convolution layer, and vgg19  loads the pretraining weights. So I guess, 1.3bit is normal? 
3. The resolution is 320x240 pixel which is downsampled by a factor of 2 from original image.

Thank you very much for your patience~ The most important thing is that the model can make correct predictions LOL

**Comment by mtangemann** [comment]
1. It is `(log_likelihood - log_likelihood_uniform) * log2(e)`. As I mentioned in my post before, the negative log likelihood you optimize during training is not the number you would report in a paper. In the example you gave, the NLL of 11.5 would amount to 1.6bit (not 1.3, sorry I made an error there before) and the 10.0 at the end of the training are roughly 3.8bit information gain (assuming 640x480px).

3. Then the negative log likelihood of the uniform distribution changes, and therefore also the information gain. With 320x240px you would have roughly -0.4bit information gain at the beginning and 1.8bit information gain at the end of the training. That looks perfectly reasonable.

2. With the correct resolution, the information gain before training (-0.4bit) looks perfectly fine to me.

**Comment by nicholast1985** [comment]
Shouldn't we compute the log-likehood for the centerbias for each frame to get a metric for how good the centerbias is for each frame, and use those values to subtract from the model's log-likelihood.
We'd essentially treat the center bias as a model, compute the log-likelihood using `cb_log_likelihood = torch.mean(torch.sum(fixation * log_cb.double(), dim=(1,2)) / torch.sum(fixation, dim=(1,2)))` for each frame, and then compute the information gain as a difference between `model_log_likelihood` and `cb_log_likelihood`.
